Updated CONTRIBUTING.md (and moved Mise instructions to it). (#4699)
wasp-lang/wasp: CONTRIBUTING.md

# Contributing to Wasp

There are several main ways in which you can contribute to Wasp:

- [Wasp as a web framework](#wasp-as-a-web-framework) (React, Node, HTML/CSS, database and so on)
- [Wasp CLI/compiler internals](#wasp-clicompiler-internals) (Haskell)
- [Tutorials or Example apps](#tutorials-or-example-apps)
- [Documentation](#documentation)

## Before you begin

- Get some experience with using Wasp, if you don't have any yet.

>   Check out the [**Getting Started**](https://wasp.sh/docs) guide to get familiar with Wasp's fundamentals.

Reviewer: I don't think this is rendering as multiple lines on GitHub, if that's your intention

Author: This was just so it is nicer to write in editor, I add two spaces at the end when I want multiple lines to actually render.
Who knows what I wanted here though I might check!

Author: Ok yeah I will let it be as it is, single line rendering is fine.

  Ideally, you'd also build an app from the [**Pick a Tutorial**](https://wasp.sh/docs/tutorials/todo-app) page to really get a feel for it!

> - Figure out what you'd like to help with. It can be code, documentation, tutorials, etc. More on this below.

Reviewer: Maybe we can add a specific note like

```md
- Ping us on the GitHub Issue or Discord #wasp-dev channel with the task you want to do, and the general solution you have thought of. This will reduce duplicate efforts and later make PR review way faster.
```

Author: It makes sense! I will try to fit it in.
The main reason why I decided not to go too much into explaining is that I think it dooesn't really help -> those who want to put the effort will do it anyway, those who won't won't care. So I decided helping them too much doesn't really do anything, maybe just lets non-honest contributors get further without red flags.
But I might add this one.

- Join our Discord [![**Discord**](https://img.shields.io/discord/686873244791210014?label=chat%20on%20discord)](https://discord.gg/rzdnErX) for faster communication and feedback. We'd be happy to help you find the issue you'll enjoy working on, depending on your interests and skill set!

## This repo

We are using a monorepo approach, where one git repo contains multiple related projects.

> In our case, those are [web](web/) (Wasp's web page, including docs and blog), [waspc](waspc/) (Wasp CLI & framework), etc.

Reviewer: I'd mention the examples and wasp-app-runner too, they seem notable

Author: I put etc. so we don't have to update this much. I would skip wasp-app-runner because that one feels more likely to change / go, but I will add exapmles that is a good idea.

We use [mise](https://mise.jdx.dev/) to manage our development tools (e.g. Haskell, Node, and code formatters). Mise is an all-in-one tool that makes it easy to set up and manage all the different tools needed for the Wasp repo. Everything is declared in a single file ([`mise.toml`](../mise.toml)), and every developer can use it to set up their environment in a consistent way. We also use it on our CI to ensure it uses the same versions of tools as well.

Run `mise install` from the root of the repo to install all the required tools. Then, you can access the mise-managed tools in different ways:

- **(Recommended for local development)** You can set up your shell to automatically call the `mise activate` script. This will make sure that the specified tools and versions are in your `PATH` when you go into the repo. Check their installation instructions at https://mise.jdx.dev/installing-mise.html#shells.

> - You can also run [`mise en`](https://mise.jdx.dev/cli/en.html) to go into an one-off shell for the current project, similar to `nix-shell` or `virtualenv`.

Reviewer: ```suggestion
- You can also run [`mise en`](https://mise.jdx.dev/cli/en.html) to go into a one-off shell for the current project, similar to `nix-shell` or `virtualenv`.
```

- For one-off commands, you can use [the `mise exec` command](https://mise.jdx.dev/cli/exec.html) (or `mise x`) to run a specific command with the repo tools available, e.g. `mise x -- ghc --version`, `mise x -- node --version`, `mise x -- ./run build`, etc.

You can learn more and install Mise by following the [official instructions](https://mise.jdx.dev/getting-started.html), then run `mise install` from the repo root to install the required tools.

> [!NOTE]
> There are no hard dependencies on mise for local development, so if you prefer to use your own tooling, you can install each program separately, and use the versions specified in [`mise.toml`](mise.toml) as a reference. But then, you're in charge of making sure you have the right versions of the tools installed, and keeping them up-to-date as we upgrade them.

### Basic commands

Formatting is defined at the repo level, and you can run it from the repo root with
```

>   npm run format:check

Reviewer: why the extra space?

Author: My old habit, will remove.

```
to check the formatting without any changes to the files, or
```
  npm run format:write
```
to check and automatically fix formatting (modifies the files).

## Ways to contribute

### Wasp as a web framework

Wasp is a language for developing full-stack web apps. This means there are plenty of tasks related to web development itself, and most come down to improving one of the core features of Wasp (like Auth, Operations, Jobs, Rendering), or maybe improving our dev tooling config, etc.

If you have full-stack experience outside of Wasp, you should already be in a good place to start contributing in this direction.

[**Web dev issues for beginners can be found here.**](https://github.com/wasp-lang/wasp/issues?q=is%3Aopen+is%3Aissue+label%3A%22good+first+issue%22+label%3Awebdev)

### Wasp CLI/compiler internals

> Wasp compiler is implemented in Haskell, but you will also see a lot of Javascript and other web technologies because Wasp compiles it's own code into them.

Reviewer: ```suggestion
The Wasp compiler is implemented in Haskell, but you will also see a lot of Javascript and other web technologies because Wasp compiles its own code into them.
```

> While you will need to know some Haskell, you don't have to be an expert in Haskell to contribute or understand the code since we don't use complicated Haskell features much -> most of the code is relatively straightforward, and we are happy to help with the part that is not.

Reviewer: A bit weirdly phrased.

```suggestion
While you will need to know some Haskell, you don't have to be an expert in Haskell to contribute or understand the code since we don't use complicated Haskell features often. Most of the code is relatively straightforward, and we are happy to help with the part that is not.
```

Author: All of this copy is old, I just copied it. But happy to apply these improvements, will do!
